Add MPS writer component

hanno-arm opened the pull request:
__Context:__ This PR is the second in a series aiming to upstream the 'MPS' (Message Processing Stack) component that is currently being developed alongside the TLS 1.3 prototype implementation in https://github.com/hannestschofenig/mbedtls/tree/tls13-prototype. The first PR was #4014 which is still unmerged and on which this PR is based. The goal is to upstream the components of MPS separately and with unit tests, but not yet put to use from the handshake layer (with the exception of the MPS reader, as explained in #4014) Once all components of MPS and the TLS 1.3 implementation have been upstreamed, MPS will replace the current messaging layer of Mbed TLS.

__This PR:__ This PR adds the 'writer' component of MPS.

__Dependencies:__ Based on #4014 which needs to be merged first.

Please see the header files for extensive documentation of the reader and the API.

hanno-arm commented:
Rebased to resolve conflicts.

ronald-cron-arm reviewed (changes requested):
I've started the review but did not complete it. I've reviewed mps_writer.h and a third of mps_writer.c. I will continue tomorrow. Otherwise I've checked the test coverage and it is not 100% thus please add the necessary test cases.

hanno-arm commented on library/mps_error.h:
```diff
@@ -100,4 +100,39 @@
+#define MBEDTLS_MPS_WRITER_MAKE_ERROR(code) \
+    ( -( MBEDTLS_MPS_WRITER_ERR_BASE | (code) ) )
+
+/*
```
Done.

hanno-arm commented on library/mps_error.h:
```diff
@@ -100,4 +100,39 @@
+
+/** An attempt was made to reclaim a buffer from the writer,
+ *  but the buffer hasn't been fully used up, yet.            */
+#define MBEDTLS_ERR_MPS_WRITER_DATA_LEFT             MBEDTLS_MPS_WRITER_MAKE_ERROR( 0x1 )
```
This is common for all error codes in the library, so I'd rather not touch this now. See e.g. ssl.h.

hanno-arm commented on library/mps_writer.c:
```diff
@@ -0,0 +1,424 @@
+                                .queue_next      = 0,
+                                .queue_remaining = 0 };
+
+    *wr = dst;
```
Thanks, I didn't know that. Since it doesn't make a difference to the compiler or readability, I prefer not to try out whether VST likes it or not, and leave it as is.

hanno-arm commented on library/mps_writer.c:
```diff
@@ -0,0 +1,424 @@
+                             mbedtls_mps_size_t buf_len )
+{
+    unsigned char *queue;
+    mbedtls_mps_size_t copy_from_queue;
```
I prefer keeping the initialization of the variable to the place where it's actually used, which makes reasoning about correctness more local. 

Do you have a specific reason why you'd like to move the initialization to here?

hanno-arm commented on library/mps_writer.c:
```diff
@@ -0,0 +1,424 @@
+    unsigned char *queue;
+    mbedtls_mps_size_t copy_from_queue;
+    MBEDTLS_MPS_TRACE_INIT( "writer_feed, buflen %u",
+                (unsigned) buf_len );
```
Fixing.

hanno-arm commented on library/mps_writer.c:
```diff
@@ -0,0 +1,424 @@
+    {
+        mbedtls_mps_size_t queue_next, queue_remaining;
+        queue_remaining = wr->queue_remaining;
+        queue_next = wr->queue_next;
```
Fine by me, fixing.

hanno-arm commented on library/mps_writer.c:
```diff
@@ -0,0 +1,424 @@
+        copy_from_queue = MIN( queue_remaining, buf_len );
+        queue += queue_next;
+
+        if( copy_from_queue != 0 )
```
Fine by me - changing.

hanno-arm commented on library/mps_writer.c:
```diff
@@ -0,0 +1,424 @@
+
+        /* Check if, after the last copy, the entire
+         * queue has been dispatched. */
+        queue_remaining -= copy_from_queue;
```
Changing.

hanno-arm commented on library/mps_writer.c:
```diff
@@ -0,0 +1,424 @@
+    MBEDTLS_MPS_TRACE_RETURN( 0 );
+}
+
+static int mps_writer_fragment_committed( mbedtls_mps_writer *wr )
```
Ok

hanno-arm commented on library/mps_writer.c:
```diff
@@ -0,0 +1,424 @@
+    return( committed >= out_len );
+}
+
+static int mps_writer_committed_data_in_queue( mbedtls_mps_writer *wr )
```
Changed to `mps_writer_some_data_in_queue_is_committed()`.

hanno-arm commented on library/mps_writer.c:
```diff
@@ -0,0 +1,424 @@
+    return( commit > out_len - overlap );
+}
+
+static void mps_writer_copy_queue_to_fragment( mbedtls_mps_writer *wr )
```
Changing.

hanno-arm commented on library/mps_writer.c:
```diff
@@ -0,0 +1,424 @@
+    queue_size = commit - ( out_len - queue_overlap );
+    copy_from_queue = MIN( queue_overlap, queue_size );
+
+    if( copy_from_queue != 0 )
```
Changing.

hanno-arm commented on library/mps_writer.h:
```diff
@@ -0,0 +1,377 @@
+ * 'queue' buffer to hold the data to the consumer. The contents of this queue
+ * buffer will be gradually split among the next outgoing data buffers when
+ * the provider subsequently provides them; in the example, the CRT chain would
+ * be split amont multiple records when the record layer hands more plaintext
```
Fixing.

hanno-arm commented on library/mps_writer.h:
```diff
@@ -0,0 +1,377 @@
+#include "mps_common.h"
+#include "mps_error.h"
+
+struct mbedtls_mps_writer;
```
Actually, this isn't necessary at all, somewhat surprisingly. Removing this line, then, as this is also how the `typedef`s in `ssl.h` are placed, prior to the actual struct definitions.

hanno-arm commented on library/mps_writer.h:
```diff
@@ -0,0 +1,377 @@
+struct mbedtls_mps_writer;
+typedef struct mbedtls_mps_writer mbedtls_mps_writer;
+
+/** \brief The type of states for the reader.
```
Fixing.

hanno-arm commented on library/mps_writer.h:
```diff
@@ -0,0 +1,377 @@
+ *
+ *  Possible values are:
+ *  - #MBEDTLS_MPS_WRITER_PROVIDING (initial state)
+ *    The writer awaits buffers for holding outgoing
```
Agree - changing.

hanno-arm commented on library/mps_writer.h:
```diff
@@ -0,0 +1,377 @@
+ *    The writer awaits buffers for holding outgoing
+ *    data to be assigned to it via mbedtls_writer_feed().
+ *  - #MBEDTLS_MPS_WRITER_CONSUMING
+ *    The writer has buffers to serve write requests from.
```
Changing.

hanno-arm commented on library/mps_writer.h:
```diff
@@ -0,0 +1,377 @@
+ *  - #MBEDTLS_MPS_WRITER_CONSUMING
+ *    The writer has buffers to serve write requests from.
+ **/
+typedef unsigned char mbedtls_mps_writer_state_t;
```
Hmm, good point. I somehow thought that we're avoiding enums, thinking e.g. of the list of ciphersuite constants in `ssl_ciphersuites.h` which are't in a macro. But those are actually standard defined constants.

Changing this to an enum, then.

hanno-arm replied:
Ah, now I remember I think... I wanted control over the size of the enum, which C99 doesn't give me. Some compilers allow to specify e.g. `-fshort-enums`, but I didn't want to depend on that.

Hmm... leaving it for now.

hanno-arm commented on library/mps_writer.h:
```diff
@@ -0,0 +1,377 @@
+
+struct mbedtls_mps_writer
+{
+    /** The current buffer to hold outgoing data.      */
```
Changing.

hanno-arm commented on library/mps_writer.h:
```diff
@@ -0,0 +1,377 @@
+    unsigned char *queue;
+    /** The size in bytes of the outgoing data buffer \c out. */
+    mbedtls_mps_stored_size_t out_len;
+    /** The length of the queue buffer \c queue. */
```
Changing.

hanno-arm commented on library/mps_writer.h:
```diff
@@ -0,0 +1,377 @@
+     */
+    mbedtls_mps_stored_size_t committed;
+
+    /** The offset from the beginning of the outgoing data buffer of the
```
Changing.

hanno-arm commented on library/mps_writer.h:
```diff
@@ -0,0 +1,377 @@
+     */
+    mbedtls_mps_stored_size_t end;
+
+    /** In consuming state, this denotes the size of the overlap between the
```
Similar to the reader, the meaning of this field depends on whether the writer is producing or consuming.

In producing, it's the amount of data copied into the queue buffer.

In consuming, it is indeed the amount of overlap between queue and output buffer, or 0 if no queue is used.

In the reader, I used the `acc_share` union for the overlap. I suggest to do the same here, as I agree that unconditionally naming it `queue_next` (but loading it as `queue_overlap` in many places) is confusing.

Ok?

ronald-cron-arm replied:
Yes with the union we can have two names which will make things easier to understand. Otherwise, in the case of the CONSUMING state wouldn't it be better to keep in the context not the overlap but rather the actual used size of the output buffer. Thus rather keeping `outlen - overlap` rather than the overlap. Looking at the code it seems that but in one case the overlap is always used as`outlen-overlap`. To me it would also make things easier to reason about.

hanno-arm commented on library/mps_writer.h:
```diff
@@ -0,0 +1,377 @@
+     *  queue and the current out buffer, once <code>end > out_len</code>.
+     *  If <code>end < out_len</code>, its value is \c 0.
+     *  In providing state, this denotes the amount of data from the queue that
+     *  has already been copied to some outgoing data buffer.
```
The documentation isn't accurate anymore, for a slightly subtle reason: Initially, the writer decides to use the queue only if you request data past the end of the output buffer. At this point, we'll thus indeed have `end > out_len`. However, if the consumer then calls `commit_partial(...)` which re-sets the 'marker' to somewhere within the overlap of queue and output buffer, we maintain the value of `queue_next`, remembering the overlap, and we're continuing to serve from the queue even if subsequent requests happen to be fully within the output buffer.

I reckon this looks odd, but here's the rationale -- maybe you have suggestions for improvement:
* Firstly, the raison d'etre for `commit_partial()` is to not force handshake message writing functions to pre-compute the total length of the handshake message. Ultimately, this should be possible, but partly cumbersome -- e.g. either or both of our ECDHE or ECDSA API don't provide information on how large their output will be, IIRC -- so the goal is to start by allowing message writers to request the biggest buffer possible, do their writing, and then say how much they've actually used, which is what `commit_partial()` allows them to do. 
* The above could then naturally happen as follows: We're running a handshake over a low-MTU network, thus using small records, but have a larger queue buffer to be able to temporarily hold handshake messages such as the certificate which exceed the MTU. Message writers that don't know the size of the handshake message upfront would then acquire and write to the queue, but if it happens that the message turns out to be very small and fitting into the low-size records adjusted to the MTU, then the `commit_partial()` runs into the situation described above.
* The previous implementation of `commit_partial()` immediately copied the overlap between queue and output buffer to the output buffer. In this case, it was OK to re-set the queue overlap to `0` and continue as if we had never used the queue. However, in alignment with the reader, and considerably removing complexity in later layers of MPS, I changed this in the latest commits, to have `commit()` and `commit_partial()` just set a marker, and leave all copying to `reclaim()`. With this model, I have to remember that I have already used the queue in the situation where a `commit_partial()` resets the marker back into the output buffer.

Does that make sense? I hope it sheds a bit of light on the thinking behind the interface and implementation.

hanno-arm replied:
I updated the documentation, removing the claim that `queue_next` is only non-zero of `end > out_len`, which as explained isn't true.

hanno-arm commented on library/mps_writer.h:
```diff
@@ -0,0 +1,377 @@
+ * \param writer    The writer context to be used.
+ * \param buf       The buffer that outgoing data can be written to
+ *                  and that the writer should manage.
+ * \param buflen    The length of the outgoing data buffer.
```
Fixing both.

hanno-arm commented on library/mps_writer.h:
```diff
@@ -0,0 +1,377 @@
+ *                  and that the writer should manage.
+ * \param buflen    The length of the outgoing data buffer.
+ *
+ * \return          \c 0 on success. In this case, the writer is
```
I am aligning with the SSL documentation here. I'd prefer not to change this unless we decide to do it everywhere.

hanno-arm commented on library/mps_writer.h:
```diff
@@ -0,0 +1,377 @@
+ *                  from consuming to providing state.
+ *
+ * \param writer    The writer context to be used.
+ * \param queued    The address at which to store the amount of
```
Here, too, I'd prefer to bring this up in the team and fix it throughout if we think we should make use of this Doxygen feature.

ronald-cron-arm replied:
If this is the way of doing it in SSL/TLS code then yes this needs to be addressed globally.

hanno-arm commented on library/mps_writer.h:
```diff
@@ -0,0 +1,377 @@
+ *                  the state of the writer is unspecified and it must
+ *                  not be used anymore.
+ *
+ *                  On success, \c *queued contains the number of bytes that
```
Good spot, I'll add this.

hanno-arm commented on library/mps_writer.h:
```diff
@@ -0,0 +1,377 @@
+ */
+int mbedtls_mps_writer_reclaim( mbedtls_mps_writer *writer,
+                                mbedtls_mps_size_t *queued,
+                                mbedtls_mps_size_t *written,
```
Fixed.

hanno-arm commented on library/mps_writer.h:
```diff
@@ -0,0 +1,377 @@
+
+/**
+ * \brief           Check how many bytes have already been written
+ *                  to the current output buffer.
```
Thanks for spotting. I have checked and this function isn't actually used anywhere - my guess is that it has been superseded by the `written` pointer to `mbedtls_mps_writer_reclaim()`.

Until we know better, I'm removing this function.

hanno-arm commented on library/mps_writer.c:
```diff
@@ -0,0 +1,424 @@
+        /* NOTE: Currently this returns success if the provided output
+         *       buffer is exactly as big as the remaining queue,
+         *       in which case there is no space left after the
+         *       queue has been copied. Is that intentional? */
```
Yes, definitely. From the top of my head, I don't remember any tangible difference between the two behaviors, but I need to look closer.

hanno-arm commented:
@ronald-cron-arm Thanks a lot for your first review pass.

I believe to have addressed all comments except for the request for 100% test coverage, and the request to resolve a TODO in the documentation asking about whether a corner case behavior of `mbedtls_mps_writer_feed()` is intentional. I will address both at a later stage.

ronald-cron-arm commented on library/mps_writer.c:
```diff
@@ -0,0 +1,410 @@
+    if( !mps_writer_some_data_in_queue_is_committed( wr ) )
+        return;
+
+    commit = wr->committed;
```
```suggestion
    committed = wr->committed;
```

ronald-cron-arm commented:
@hanno-arm What do you think about https://github.com/ARMmbed/mbedtls/pull/4201#discussion_r612206673 ?

hanno-arm commented:
@ronald-cron-arm Good idea! I will try it and see if I run into any issues.

hanno-arm commented on library/mps_writer.h:
```diff
@@ -0,0 +1,374 @@
+ * \param writer    The writer context to be used.
+ * \param written   The address at which to store the amount of
+ *                  outgoing data that has been written to the output
+ *                  buffer last passed to mbedtls_mps_writer_feed().
```
I think the present version is correct. Could you elaborate why you think it isn't?

ronald-cron-arm commented on library/mps_writer.c:
```diff
@@ -0,0 +1,410 @@
+
+static int mps_writer_some_data_in_queue_is_committed( mbedtls_mps_writer *wr )
+{
+    mbedtls_mps_size_t const commit  = wr->committed;
```
```suggestion
    mbedtls_mps_size_t const committed  = wr->committed;
```

hanno-arm replied:
Changing.

ronald-cron-arm commented on library/mps_writer.c:
```diff
@@ -0,0 +1,410 @@
+        wr->state == MBEDTLS_MPS_WRITER_CONSUMING,
+        "Attempt to request write-buffer outside consuming mode." );
+
+    commit        = wr->committed;
```
```suggestion
    committed        = wr->committed;
```

hanno-arm replied:
Changing.

ronald-cron-arm reviewed (changes requested):
I've finished the first round of review of mps_writer.c. For the first round of reviews, I am left with the tests. I will continue tomorrow. Please find my comments below, the main one being that I have the feeling it is possible to reduce code duplication in `mbedtls_mps_writer_get()`.

hanno-arm commented on library/mps_writer.h:
```diff
@@ -0,0 +1,374 @@
+#define MBEDTLS_MPS_WRITER_RECLAIM_NO_FORCE 0
+
+/**
+ * \brief           Attempt to reclaim output buffer from writer,
```
Fixing.

hanno-arm commented on library/mps_writer.c:
```diff
@@ -0,0 +1,410 @@
+        wr->state == MBEDTLS_MPS_WRITER_CONSUMING,
+        "Can't reclaim output buffer outside of consuming mode." );
+
+    commit = wr->committed;
```
Fixing.

hanno-arm commented on library/mps_writer.c:
```diff
@@ -0,0 +1,410 @@
+    return( committed >= out_len );
+}
+
+static int mps_writer_some_data_in_queue_is_committed( mbedtls_mps_writer *wr )
```
Agree, it's trivial, but on the other hand wrapping it in a helper helps conveying the meaning of the check, so I'd prefer to stick with it.

hanno-arm commented on library/mps_writer.h:
```diff
@@ -0,0 +1,374 @@
+ *                  the writer is unmodified and can still be used.
+ *                  In particular, the writer stays in providing state.
+ * \return          #MBEDTLS_ERR_MPS_WRITER_DATA_LEFT if there is space
+ *                  left to be written in the output buffer.
```
Agreed.

hanno-arm commented on library/mps_writer.c:
```diff
@@ -0,0 +1,410 @@
+
+        wr->queue_next = 0;
+
+        if( force == 0 )
```
Fixing.

hanno-arm commented on library/mps_writer.c:
```diff
@@ -0,0 +1,410 @@
+        wr->queue_next = 0;
+
+        if( force == 0 )
+            MBEDTLS_MPS_TRACE_RETURN( MBEDTLS_ERR_MPS_WRITER_DATA_LEFT );
```
Yes, absolutely, this is deliberate. However, I realize that it isn't documented properly, which I'll fix.

hanno-arm replied:
The rationale is as follows: To the 'consumer' (e.g. the handshake message writer), it doesn't matter whether the `reclaim()` succeeded and a new buffer was fed (concretely: if the current record was full and needed to be dispatched), or the `reclaim()` failed and the writer was returned immediately for further use (concretely: there was more space left in the current record) -- the state is the same in either case: All uncommitted data has been squashed and needs to be re-written.

hanno-arm commented on library/mps_writer.c:
```diff
@@ -0,0 +1,410 @@
+    if( copy_from_queue > 0 )
+    {
+        out += out_len - queue_overlap;
+        memcpy( out, queue, copy_from_queue );
```
We could, but I don't see any benefit of doing it. Do you?

hanno-arm commented on library/mps_writer.c:
```diff
@@ -0,0 +1,410 @@
+    MBEDTLS_MPS_TRACE_RETURN( 0 );
+}
+
+static int mps_writer_queue_in_use( mbedtls_mps_writer *wr )
```
Changing.

hanno-arm commented on library/mps_writer.c:
```diff
@@ -0,0 +1,410 @@
+    MBEDTLS_MPS_TRACE_INIT( "writer_get, desired %u", (unsigned) desired );
+
+    MBEDTLS_MPS_STATE_VALIDATE_RAW( wr->state == MBEDTLS_MPS_WRITER_CONSUMING,
+                  "Attempt to request write-buffer outside consuming mode." );
```
Changed.

hanno-arm commented on library/mps_writer.c:
```diff
@@ -0,0 +1,410 @@
+                MBEDTLS_MPS_TRACE_RETURN( MBEDTLS_ERR_MPS_WRITER_OUT_OF_DATA );
+            }
+
+            /* Queue large enough, transition to serving from queue. */
```
Good spot, I'll improve on that.

hanno-arm replied:
Your outline worked well, I think the code is cleaner now.

hanno-arm commented:
Thanks @ronald-cron-arm, I have addressed your latest comments.

The test coverage improvement and documentation of corner case is still outstanding as before.

hanno-arm commented:
@ronald-cron-arm @paul-elliott-arm Please let me know when would be a good time to rebase to resolve the conflict in the VS2010 project file.

ronald-cron-arm reviewed (changes requested):
I've done the review of the tests now. I've also been through my previous comments, resolved the discussions that needed to be resolved and pointed to the pending ones.
